presto-main/src/main/java/com/facebook/presto/operator/MergeSortProcessor.java

```diff
@@ -0,0 +1,183 @@
+    {
+        this.comparator = requireNonNull(comparator, "comparator is null");
+        requireNonNull(pageSuppliers, "pageSuppliers is null");
+        ImmutableList.Builder<MergeSource> mergeSources = ImmutableList.builder();
+        for (PageSupplier pageSupplier : pageSuppliers) {
+            mergeSources.add(new MergeSource(pageSupplier));
+        }
+        this.mergeSources = mergeSources.build();
+    }
+
+    @Nullable
+    public PageWithPosition poolNextRow()
+    {
+        PageWithPosition result = null;
+        for (MergeSource mergeSource : mergeSources) {
+            PageWithPosition current = mergeSource.getPage();
+            if (current == null) {
+                if (mergeSource.isFinished()) {
+                    continue;
+                }
+                else {
+                    return null;
+                }
+            }
+            if (result == null) {
+                result = current;
+            }
+            else {
+                int compareResult = comparator.compareTo(current.getPage(), current.getPosition(), result.getPage(), result.getPosition());
+                if (compareResult < 0) {
+                    result = current;
+                }
+            }
+        }
+        return result;
+    }
+
+    public ListenableFuture<?> isBlocked()
+    {
+        if (mergeSources.stream().allMatch(source -> source.isBlocked().isDone())) {
+            return NOT_BLOCKED;
+        }
+        if (blocked != null && !blocked.isDone()) {
+            return blocked;
+        }
+        blocked = Futures.allAsList(mergeSources.stream().map(MergeSource::isBlocked).collect(Collectors.toList()));
+        return blocked;
+    }
+
+    public boolean isFinished()
+    {
+        return mergeSources.stream().allMatch(MergeSource::isFinished);
+    }
+
+    private static class MergeSource
+    {
+        private final PageSupplier pageSupplier;
+
+        private PageWithPosition currentPage;
+        private ListenableFuture<?> blocked;
+
+        public MergeSource(PageSupplier pageSupplier)
+        {
+            this.pageSupplier = requireNonNull(pageSupplier, "location is null");
+        }
+
+        public ListenableFuture<?> isBlocked()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return NOT_BLOCKED;
+            }
+            if (blocked != null && !blocked.isDone()) {
+                return blocked;
+            }
+            blocked = pageSupplier.isBlocked();
+            return blocked;
+        }
+
+        public boolean isFinished()
+        {
+            return pageSupplier.isFinished() && (currentPage == null || currentPage.isFinished());
+        }
+
+        @Nullable
+        public PageWithPosition getPage()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return currentPage;
+            }
+
+            checkState(blocked != null, "blocked is null");
```

Comment:
No, it is not. You must check whether you can call `getPage` by calling `isBlocked` first.